src/drivers/i2c/adapters/imx_i2c/imx_i2c.c

```diff
@@ -0,0 +1,223 @@
+			REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, temp);
+			log_error("arbitration lost");
+			return -EAGAIN;
+		}
+		if (for_set && (temp & mask) ) {
+			return 0;
+		}
+		if (!for_set && !(temp & mask) ) {
+			return 0;
+		}
+	}
+	return -ETIMEDOUT;
+}
+
+static int imx_i2c_tx_byte(struct imx_i2c *adapter, uint8_t byte) {
+	int res;
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2DR, byte);
+	res = imx_i2c_wait_status(adapter, IMX_I2C_I2SR_IIF, 1);
+	if (res) {
+		return res;
+	}
+	if (REG8_LOAD(adapter->base_addr + IMX_I2C_I2SR) &  IMX_I2C_I2SR_RXAK) {
+		return -ENODEV;
+	}
+	return 0;
+}
+
+static int imx_i2c_tx(struct imx_i2c *adapter, uint16_t addr, uint8_t *buff, size_t sz) {
+	int res;
+	int i;
+
+	res = imx_i2c_tx_byte(adapter, (addr << 1) & 0xFF);
+	if (res < 0) {
+		return res;
+	}
+
+	log_debug("ACK received 0x%x", addr);
+	res = sz;
+	for (i = 0; i < sz; i++) {
+		log_debug("write byte: %d=0x%X", i, buff[i]);
+		res = imx_i2c_tx_byte(adapter, buff[i]);
+		if (res < 0) {
+			return res;
+		}
+		log_debug("byte sent addr (0x%x) val (0x%x)", addr, buff[i]);
+	}
+
+	return res;
+}
+
+static int imx_i2c_stop(struct imx_i2c *adapter) {
+	uint8_t tmp;
+
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp &= ~(IMX_I2C_I2CR_MTX | IMX_I2C_I2CR_TXAK);
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	/* delay(1); */
+	tmp &= ~(IMX_I2C_I2CR_MSTA) ;
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+	usleep(100);
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, 0);
+	usleep(100);
+	return 0;
+}
+
+static int imx_i2c_start(struct imx_i2c *adapter) {
+	uint8_t tmp;
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0 );
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, IMX_I2C_I2CR_IEN );
+
+	usleep(200);
+
+	/* Start I2C transaction */
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp |= IMX_I2C_I2CR_MSTA ;
+	REG8_STORE(adapter->base_addr+ IMX_I2C_I2CR, tmp);
+
+	tmp = imx_i2c_wait_status(adapter, IMX_I2C_I2SR_IBB, 1);
+	if (tmp) {
+		return tmp;
+	}
+
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp |= IMX_I2C_I2CR_MTX | IMX_I2C_I2CR_TXAK;
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+
+	return 0;
+}
+
+
+static int imx_i2c_rx(struct imx_i2c *adapter, uint16_t addr, uint8_t *buff, size_t sz) {
+	int res;
+	int cnt;
+	uint8_t tmp;
+
+	res = imx_i2c_tx_byte(adapter, ((addr << 1) | 0x1) & 0xFF);
+	if (res < 0) {
+		return res;
+	}
+	log_debug("ACK received 0x%x", addr);
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp &= ~IMX_I2C_I2CR_MTX;
+	if (sz > 1) {
+		tmp &= ~IMX_I2C_I2CR_TXAK;
+	}
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+	/* dummy read */
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2DR);
+
+	for (cnt = sz; cnt > 0; cnt--) {
+		res = imx_i2c_wait_status(adapter, IMX_I2C_I2SR_IIF, 1);
+		if (res) {
+			log_error("i2c completion error %d", res);
```

Comment:
Put `return res;` here to remove `if (res) {return res; }` at the end of for-loop.